Unify semicolon and comment handling in nested native queries

lbrdnk opened the pull request:
Closes #29168

### Description

Prior to this PR comments were removed only from _native source queries of mbql queries_ in `fetch-source-query/source-query`. Semicolons were removed by `sql.qp/make-nestable-sql` and `sql.qp.deprecated/SQLSourceQuery`'s `to-sql` implementation. Semicolon removal was handled also for _native card references in native queries_. 

This PR unifies the logic for comment and semicolon removal in `sql.qp/make-nestable-sql`, which results in comments being also removed from _native card references in native queries_.

Drivers of both honeysql versions were considered and the tests were updated accordingly.

- `card-id-native-source-queries-test` was adjusted to run with any driver with `:sql` parent. This verifies, that _native source queries of mbql queries_ are handled correctly with drivers of both honeysql versions.
- Testcase from `card-id->source-query-and-metadata-test` was merged into the `substitute-referenced-card-query-test`. This covers comment and semicolon removal from _native card references in native queries_.

### How to verify

See #29168 for reproduction or run at least following tests with honeysql 1 and 2 driver:
- `card-id-native-source-queries-test`,
- `substitute-referenced-card-query-test`.

Change was tested locally with: `DRIVERS=h2,sqlserver,presto-jdbc clj -X:dev:ee:ee-dev:drivers:drivers-dev:test`.

### Checklist

- [X] Tests have been added/updated to cover changes in this PR.

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/metabase/metabase/30677)
<!-- Reviewable:end -->

lbrdnk committed:
Unify semicolon and comment handling in nested native queries

lbrdnk commented:
Would you please run the workflows and/or review @metamben @snoe? Thank you.

snoe commented on src/metabase/driver/sql/query_processor.clj:
```diff
@@ -44,9 +44,11 @@
   [sql]
-  (str "(" (str/replace sql #";[\s;]*$" "") ")"))
+  (str "(" (-> sql
+               (str/replace #"--.*(\n|$)" "")
```
Will this remove something like `select 'not -- a comment' from x` ? 
It would feel safer to check something like `^\s*--.*(\n|$)` but I'm not sure that fully captures the problem. 
I think `select a from x; -- comment prevents semicolon removal` is the gist right?
In that case, I think we're out of regex into sql parsing territory. I wonder if a better error message that checks for `--` in a `catch` if a query fails with a syntax error would be sufficient to point people in the right direction until we tackle parsing.

lbrdnk replied:
> Will this remove something like `select 'not -- a comment' from x` ?

I've tweaked the `make-nestable-sql` so it's bit more robust, but still it does just a regex replacement. Please check the code.

> In that case, I think we're out of regex into sql parsing territory.

Probably. Another approach how to tackle this could be to write a tokenizer that handles string literals, identifiers, semicolons, comments - which could be simpler than parsing. But making that work across all supported sql dialects would require addition of driver specific rulesets eg. as driver methods.

Mock with hardcoded "best effort" ruleset:

```clojure
(defn- tokenize-sql
    "Build vector of tokens from sql string for purposes of removal traling comments and semicolons."
    [sql]
    (loop [acc []
           str sql]
      (if (empty? str)
        acc
        (let [[type token]
              (condp re-find str
                #"^'(?:''|.|\n)*?'"        :>> #(vector :quoted %)
                #"^\"(?:.|\n)*?\""         :>> #(vector :double-quoted %)
                #"^`(?:.|\n)*?`"           :>> #(vector :backticked %)
                #"^--.*"                   :>> #(vector :comment %)
                #"^;+"                     :>> #(vector :semi %)
                #"^\s+"                    :>> #(vector :space %)
                #"^(?:\n|.)+?(?=`|\"|'|--|;|$)" :>> #(vector :rest %)
                (throw (ex-info "Failed to tokenize sql string"
                                {:sql sql
                                 :remaining str
                                 :tokenized acc})))]
          (assert (not-empty token) "Failed to tokenize sql string")
          (recur (conj acc [type token]) (subs str (count token)))))))
          
(defn remove-trailing-comments-and-semis
    "Remove trailing semicolons and comments from sql string."
    [sql]
    (let [tokens (tokenize-sql sql)
          trailing-removed (->> (rseq tokens)
                                (drop-while (fn [[type]] (#{:comment :semi :space} type)))
                                reverse)]
      (->> trailing-removed
           (map (fn [[_ token]] token))
           (str/join ""))))
```


> I wonder if a better error message that checks for `--` in a `catch` if a query fails with a syntax error would be sufficient to point people in the right direction until we tackle parsing.

As the problem is subquery specific, we would have to save information like "this is a risky query" somewhere, when subquery containing `;\s*--` is encountered eg. during source query fetching or parameter substitution -- prior to `make-nestable-sql` call. And then use it when execution of the root query fails. It would not be enough to just check the root query after the execution as `;\s*--` patterns would be already removed.

snoe reviewed (approved):
Approved but wary that this will open us up to problems stripping out too much. Would be worth thinking about other ways to approach this.

metamben commented on src/metabase/driver/sql/query_processor.clj:
```diff
@@ -44,9 +44,11 @@
   [sql]
-  (str "(" (str/replace sql #";[\s;]*$" "") ")"))
+  (str "(" (-> sql
+               (str/replace #"--.*(\n|$)" "")
```
This looks dangerous. It would throw away everything after `--` even if it occurs in a string or something similar. A smaller problem is that it might paste together two tokens that only have the comment and the newline between them. (This can be fixed by replacing with `"$1"`.

lbrdnk replied:
I've changed the approach to comment/semi removal. I think it is a bit more robust now, but as @snoe pointed out proper solution would need the sql parsing. Please have a look.

lbrdnk committed:
Use different comment and semicolon removal approach

metamben commented on src/metabase/driver/sql/query_processor.clj:
```diff
@@ -44,9 +44,32 @@
+           (as-> trimmed
+                 ;; Query could potentially end with a comment.
+                 (if (re-find #"--.*$" trimmed)
+                   (str/replace trimmed #"$" "\n")
```
```suggestion
                   (str trimmed "\n")
```

metamben commented on src/metabase/driver/sql/query_processor.clj:
```diff
@@ -44,9 +44,32 @@
-  (str "(" (str/replace sql #";[\s;]*$" "") ")"))
+  (str "("
+       (-> sql
+           (str/replace #";([\s;]*(--.*\n?)*)*?$" "")
```
```suggestion
           (str/replace #";([\s;]*(--.*\n?)*)*$" "")
```

lbrdnk committed:
Apply suggestions from code review

Co-authored-by: metamben <103100869+metamben@users.noreply.github.com>

lbrdnk committed:
Update base and effective type verification for oracle

lbrdnk commented:
I've modified `nested-queries-test/breakout-results` to handle different base type and effective type for `:oralce`. Looking at eg. [`metabase.driver.sql-jdbc.sync.describe-table-test/database-types-fallback-test`](https://github.com/metabase/metabase/blob/8305374a1e7e938b4a10e27f20344ea577e4fde8/test/metabase/driver/sql_jdbc/sync/describe_table_test.clj#L101) made me decide it is ok, that oracle uses different types in some circumstances.

Then, `metabot-util-test/inner-query-name-collisions-test` that failed for `:mysql` is passing locally with: `DRIVERS=mysql clj -X:dev:ee:ee-dev:drivers:drivers-dev:test`. Hence there are either some environment differences or it is a flake.

Rest of the failures is happening because of missing secrets for ci runs on external branches.

@metamben, if you agree with the change, would you mind re-running the workflows? Thank you.

lbrdnk commented:
@metamben, all remaining failures are from tests missing secrets. If you are fine running the code with secrets, please open new branch so I can change the base of the PR, as in #29053. I've also rebased the branch to the latest master.

metamben commented:
@lbrdnk sorry, it seems we need another rebase. :(

lbrdnk commented:
@metamben, done. I've also successfully run tests with h2 mongo and oracle on my machine.

metamben reviewed (approved):
LGTM

metamben committed:
Merge branch 'master' into 29168-ending-with-comment

metamben commented:
@lbrdnk, opened #32900, when it succeeds, we force merge this PR.

lbrdnk committed:
Avoid running test with snowflake and bigquery

metamben commented on test/metabase/query_processor_test/nested_queries_test.clj:
```diff
@@ -75,19 +78,21 @@
-            (not has-source-metadata?)
-            (dissoc :id :semantic_type :settings :fingerprint :table_id :coercion_strategy))
-          (qp.test/aggregate-col :count)]})
+  ;; Oracle driver returns base and effective type as Decimal, other drivers as Integer.
+  (let [field-type (case driver/*driver* :oracle :type/Decimal :type/Integer)]
```
What's the reason this surfaces in this PR?

lbrdnk replied:
I've wanted to extend `card-id-native-source-queries-test` to query not only h2, but also other dbms.

To the best of my knowledge, when `mt/test-drivers` is not used, h2 database is queried by default, even if it is not in the set of drivers currenly tested.

So why extending the test? Because there are still some drivers using honeysql 1. And comment removal from native card query has different code path, depending on used honeysql version.

Extenting the test this way requires:
1. Getting the right database name for the native query in the test code. (Current bigquery and snowflake failures)
2. Adjusting the `breakout-results` to handle small differences between drivers.

Those being different field types for oracle, and field name formatting.

---

Actually while writing this, I've realized:
1. how to avoid this `case *driver*` snippet completely,
2. how to make the test runnable even for bigquery and snowflake (hopefully, as I have no way to verify that locally). 

I will push the update shortly.

metamben commented on test/metabase/query_processor_test/nested_queries_test.clj:
```diff
@@ -312,25 +317,31 @@
-    (is (= (breakout-results :has-source-metadata? false :native-source? true)
-           (run-native-query "SELECT * FROM VENUES -- small comment here\n"))
-        "Ensure trailing comments followed by a newline are trimmed and don't cause a wrapping SQL query to fail")))
+  (mt/test-drivers (set/intersection (set/difference (mt/normal-drivers-with-feature :nested-queries)
+                                                     #{:bigquery-cloud-sdk :snowflake})
```
You can use `disj` instead of `set/difference` to "remove" elements from a set.

metamben committed:
Merge branch 'master' into 29168-ending-with-comment

lbrdnk committed:
Update `breakout-results` and test

1. In `breakout-results` case driver statement was redundant.
Original code had field name and types hardcoded. During further
examination, I've realized name and types are available thanks
to the use of `qp.test/col`, so are just "unhardcoded" :)

2. `card-id-native-source-queries-test` is altered, so instead of
hardcoding native sub-queries, roughly equivalent* mbql query is
compiled and its native form is used. This way I can hopefully avoid
problems with extraction of cloud databases' table names for testing
purposes.

* By roughly equivalent I mean instead of using * for selecting
fields, mbql is compiled selecting fields explicitly. Also limit is added
in some cases, but local testing showed no problems with that.

lbrdnk committed:
Merge remote-tracking branch 'upstream/master' into 29168-ending-with-comment

lbrdnk commented:
@metamben, I've merged in fresh master and resolved conflicts.
